src/sources/socket/mod.rs

```diff
@@ -360,13 +444,84 @@ mod test {
         );
     }
 
+    #[test]
+    fn udp_shutdown_simple() {
+        let (tx, rx) = mpsc::channel(2);
+        let source_name = "udp_shutdown_simple";
+
+        let mut shutdown = ShutdownCoordinator::new();
+        let (address, mut rt, source_handle) =
+            init_udp_with_shutdown(tx, source_name, &mut shutdown);
+
+        send_lines_udp(address, vec!["test"]);
+        let events = rt.block_on(collect_n(rx, 1)).ok().unwrap();
+
+        assert_eq!(
+            events[0].as_log()[&event::log_schema().message_key()],
+            "test".into()
+        );
+
+        // Now signal to the Source to shut down.
+        shutdown.shutdown_source_begin(source_name);
+        let deadline = Instant::now() + Duration::from_secs(1000);
+        let shutdown_success =
+            shutdown.shutdown_source_end(&mut rt, source_name.to_string(), deadline);
+        assert_eq!(true, shutdown_success);
+
+        // Ensure source actually shut down successfully.
+        rt.block_on(source_handle).unwrap();
+    }
+
+    // #[test]
+    // fn udp_shutdown_infinite_stream() {
+    //     let (tx, rx) = mpsc::channel(1000);
+    //     let source_name = "udp";
+    //
+    //     let mut shutdown = ShutdownCoordinator::new();
+    //     let (address, mut rt, source_handle) =
+    //         init_udp_with_shutdown(tx, source_name, &mut shutdown);
+    //
+    //     // Future that keeps sending lines to the UDP source forever.
+    //     // We leak this thread and let it run forever until the test binary terminates.
+    //     let pump_handle = std::thread::spawn(move || {
+    //         send_lines_udp(address, std::iter::repeat("test"));
+    //     });
+    //
+    //     // This consumes `rx` which causes the Source to error when it tries to forward its results
+    //     // into `tx` but finds the other side of the channel has been dropped.
+    //     let events = rt.block_on(collect_n(rx, 100)).ok().unwrap();
```

Comment:
This line is the problem.  `collect_n` takes ownership of `rx` so when that function completes `rx` gets dropped.  This causes the actual UDP source future to Err when it tries to forward the events it is receiving to its copy of `tx` but `tx` errors cause its Receiver has already been closed.  It doesn't look like futures01 provides a way to get items out of a `mpsc::Receiver` without consuming it.

EDIT: Ana let me know that futures 0.3 is available now, and it looks like the [try_next()](https://docs.rs/futures/0.3.4/futures/channel/mpsc/struct.Receiver.html#method.try_next) method that was added in newer versions may do what I need.  I'm gonna take a break on this for the rest of the day to work on other stuff cause I'm burned out on working on this test, but tomorrow I might take a shot at reviving this with the newer futures library.

Reply:
> Ana let me know that futures 0.3 is available now, and it looks like the try_next() method that was added in newer versions may do what I need. I'm gonna take a break on this for the rest of the day to work on other stuff cause I'm burned out on working on this test, but tomorrow I might take a shot at reviving this with the newer futures library.

Have a nice well deserved rest. :)

Reply:
Can you have `collect_n` return `rx`? so that you can drop it later?

Reply:
Yeah, the normal pattern for this kind of thing (you'll see it a lot in the `Sink` trait) is to return a tuple of `(result, original_thing)` so that you can poll it again later.

You could also look at `CollectCurrent` in our test utils and see if you could use that instead.

Reply:
> Can you have `collect_n` return `rx`? so that you can drop it later?

I tried this, but I get the error "error[E0507]: cannot move out of `rx`, a captured variable in an `FnMut` closure".  I actually decided to ask this on [StackOverflow here](https://stackoverflow.com/questions/60645623/moving-object-into-and-then-back-out-of-fnmut-closure), if there's something stupid I'm missing please let me know!  If I don't get a good answer back on this then I'll give the newer future library a shot next and see how that goes.

Reply:
Ooh.  CollectCurrent is *almost* what I want, just needs some work to take a limit on how many items to collect and to block until that many items are indeed collected.  It does, however, demonstrate that what I want to do is possible.  I thought I learned my lesson last week but I guess I'm still not thinking with Combinators!  Gonna sign off for the night but take a crack at writing my own collect combinator to do what I want and finally get the more interesting unit tests into review tomorrow.  Thanks for the tip @lukesteensen!